# Rareté des matchs nuls

La Premier League est le championnat d'Angleterre de football où s'affrontent du mois d'août au mois de mai suivant, vingt clubs professionnels.

Les données proviennent du site http://www.football-data.co.uk/englandm.php
Nous utiliserons les données de la saison 2013-2014 jusqu'à la saison 2018-2019

Nous garderons en variable la date, le nom des équipes à domicile et à l'extérieur et leur nombre de buts marqués.

Les objectifs sont de :
- Etudier les scores obtenus pour les matchs "Everton-Liverpool" ; le nombre de matchs nuls est-il exceptionnel?
- Determinier et représenter la loi de probabilité des scores obtenus en Premier League

## Etape 1 : Matchs nuls

#### Lecture et traitement du jeu de données

In [32]:
import pandas as pd

In [33]:
df1 = pd.read_csv('./data/foot_1league_2013_2014.csv')
df2 = pd.read_csv('./data/foot_1league_2014_2015.csv')
df3 = pd.read_csv('./data/foot_1league_2015_2016.csv')
df4 = pd.read_csv('./data/foot_1league_2016_2017.csv')
df5 = pd.read_csv('./data/foot_1league_2017_2018.csv')
df6 = pd.read_csv('./data/foot_1league_2018_2019.csv')

In [35]:
T = df1.append(df2, ignore_index=True)
T = T.append(df3, ignore_index=True)
T = T.append(df4, ignore_index=True)
T = T.append(df5, ignore_index=True)
T = T.append(df6, ignore_index=True)

T.head()

Div      Date    HomeTeam     AwayTeam  FTHG  FTAG FTR  HTHG  HTAG HTR  ...  \
0  E0  17/08/13     Arsenal  Aston Villa   1.0   3.0   A   1.0   1.0   D  ...   
1  E0  17/08/13   Liverpool        Stoke   1.0   0.0   H   1.0   0.0   H  ...   
2  E0  17/08/13     Norwich      Everton   2.0   2.0   D   0.0   0.0   D  ...   
3  E0  17/08/13  Sunderland       Fulham   0.0   1.0   A   0.0   0.0   D  ...   
4  E0  17/08/13     Swansea   Man United   1.0   4.0   A   0.0   2.0   A  ...   

  BbAv<2.5  BbAH  BbAHh  BbMxAHH  BbAvAHH  BbMxAHA  BbAvAHA  PSCH  PSCD   PSCA  
0     2.37   NaN    NaN      NaN      NaN      NaN      NaN  1.44  5.00   8.05  
1     2.02   NaN    NaN      NaN      NaN      NaN      NaN  1.42  4.62  10.19  
2     1.82   NaN    NaN      NaN      NaN      NaN      NaN  3.81  3.27   2.21  
3     1.77   NaN    NaN      NaN      NaN      NaN      NaN  2.52  3.23   3.16  
4     1.98   NaN    NaN      NaN      NaN      NaN      NaN  3.62  3.41   2.22  

[5 rows x 68 columns]

In [36]:
col = ['Date', 'HomeTeam', 'AwayTeam', 'FTHG', 'FTAG']
T = T[col]

T.rename(columns={'HomeTeam' : "DOM", 'AwayTeam' : "EXT", 
                  'FTHG' : "BDOM", 'FTAG' : "BEXT"}, inplace=True)

T.head()

Date         DOM          EXT  BDOM  BEXT
0  17/08/13     Arsenal  Aston Villa   1.0   3.0
1  17/08/13   Liverpool        Stoke   1.0   0.0
2  17/08/13     Norwich      Everton   2.0   2.0
3  17/08/13  Sunderland       Fulham   0.0   1.0
4  17/08/13     Swansea   Man United   1.0   4.0

In [37]:
len(T)

2281

Nous avons donc 2281 matchs joués 

#### Fréquence de matchs nuls

In [38]:
len(T.query('BDOM == 0 and BEXT == 0')) # On regarde le nombre de 0-0

171

In [39]:
len(T.query('BDOM == 0 and BEXT == 0'))/len(T)
# On regarde la fréquence de 0-0

0.07496711968434897

- Il y a environ 7% de 0-0

In [40]:
T.query('BDOM == BEXT') # Voici tout les matchs nuls

Date           DOM           EXT  BDOM  BEXT
2       17/08/13       Norwich       Everton   2.0   2.0
12      24/08/13       Everton     West Brom   0.0   0.0
15      24/08/13     Newcastle      West Ham   0.0   0.0
16      24/08/13   Southampton    Sunderland   1.0   1.0
20      26/08/13    Man United       Chelsea   0.0   0.0
...          ...           ...           ...   ...   ...
2267  05/05/2019       Arsenal      Brighton   1.0   1.0
2269  05/05/2019  Huddersfield    Man United   1.0   1.0
2275  12/05/2019     Leicester       Chelsea   0.0   0.0
2278  12/05/2019   Southampton  Huddersfield   1.0   1.0
2279  12/05/2019     Tottenham       Everton   2.0   2.0

[532 rows x 5 columns]

In [46]:
p = len(T.query('BDOM == BEXT'))/len(T)
p

0.23323103901797457

- Il y a environ 23% de matchs nuls

## Etape 2 : Matchs joués entre Everton et Liverpool

In [42]:
S = T.query('DOM in ["Everton", "Liverpool"] and EXT in ["Everton", "Liverpool"]')
S

Date        DOM        EXT  BDOM  BEXT
111     23/11/13    Everton  Liverpool   3.0   3.0
221     28/01/14  Liverpool    Everton   4.0   0.0
434     27/09/14  Liverpool    Everton   1.0   1.0
611     07/02/15    Everton  Liverpool   0.0   0.0
839   04/10/2015    Everton  Liverpool   1.0   1.0
1096  20/04/2016  Liverpool    Everton   4.0   0.0
1310    19/12/16    Everton  Liverpool   0.0   1.0
1428    01/04/17  Liverpool    Everton   3.0   1.0
1678  10/12/2017  Liverpool    Everton   1.0   1.0
1837  07/04/2018    Everton  Liverpool   0.0   0.0
2040  02/12/2018  Liverpool    Everton   1.0   0.0
2187  03/03/2019    Everton  Liverpool   0.0   0.0

In [44]:
S.query('BDOM == BEXT') #Matchs nuls entre Everton et Liverpool

Date        DOM        EXT  BDOM  BEXT
111     23/11/13    Everton  Liverpool   3.0   3.0
434     27/09/14  Liverpool    Everton   1.0   1.0
611     07/02/15    Everton  Liverpool   0.0   0.0
839   04/10/2015    Everton  Liverpool   1.0   1.0
1678  10/12/2017  Liverpool    Everton   1.0   1.0
1837  07/04/2018    Everton  Liverpool   0.0   0.0
2187  03/03/2019    Everton  Liverpool   0.0   0.0

In [45]:
len(S.query('BDOM == BEXT'))

7

- Il y a eu 7 matchs nuls entre Everton et Liverpool en 12 rencontres

- On a p la probabilité d'obtenir un match nul (0.23) en 12 matchs
- On peut utiliser une loi binomiale selon ses paramètres.

In [47]:
from scipy.stats import binom

X = binom(12, p)
s = 0
for k in range(7,12):
    s += X.pmf(k)

s    

0.009600653238968093

- On a à peine 1% de chance d'obtenir au moins 7 matchs nuls en 12 confrontations

In [52]:
# On pouvait calculer ce pourcentage autrement.
# En effet, on a calculé P(X>=7) mais on pouvait calculer 1 - P(X<6)

# Et P(X<k) peut se calculer avec X.cdf(k)

1 - X.cdf(6)

0.00960067914710383

## Etape 3 : Distribution des scores

#### Table des scores

In [62]:
R = T.groupby(['BDOM', 'BEXT']).count()
R = R[['Date']]
R.reset_index(inplace=True)
R  # Ici on a tout les scores différents des matchs et le nombre de fois où ils sont arrivés

BDOM  BEXT  Date
0    0.0   0.0   171
1    0.0   1.0   173
2    0.0   2.0    96
3    0.0   3.0    67
4    0.0   4.0    26
5    0.0   5.0     5
6    0.0   6.0     2
7    1.0   0.0   230
8    1.0   1.0   221
9    1.0   2.0   156
10   1.0   3.0    70
11   1.0   4.0    30
12   1.0   5.0     8
13   1.0   6.0     1
14   1.0   7.0     1
15   2.0   0.0   204
16   2.0   1.0   171
17   2.0   2.0   119
18   2.0   3.0    44
19   2.0   4.0     8
20   2.0   5.0     2
21   2.0   6.0     1
22   3.0   0.0    96
23   3.0   1.0   112
24   3.0   2.0    48
25   3.0   3.0    21
26   3.0   4.0     5
27   3.0   5.0     1
28   3.0   6.0     2
29   4.0   0.0    50
30   4.0   1.0    46
31   4.0   2.0    21
32   4.0   3.0    10
33   4.0   5.0     1
34   5.0   0.0    20
35   5.0   1.0    16
36   5.0   2.0     1
37   5.0   3.0     3
38   5.0   4.0     2
39   6.0   0.0     5
40   6.0   1.0     7
41   6.0   2.0     2
42   6.0   3.0     2
43   7.0   0.0     1
44   7.0   2.0     1
45   8.0   0.0     1

- On voit qu'il y a eu 171 '0-0' et 173 '0-1'

#### Fréquence des scores

In [65]:
R['Freq'] = R['Date']/R['Date'].sum()
R.head()

BDOM  BEXT  Date      Freq
0   0.0   0.0   171  0.075000
1   0.0   1.0   173  0.075877
2   0.0   2.0    96  0.042105
3   0.0   3.0    67  0.029386
4   0.0   4.0    26  0.011404

In [69]:
R.sort_values(by='Freq', ascending=False, inplace=True)
R.head(10)

BDOM  BEXT  Date      Freq
7    1.0   0.0   230  0.100877
8    1.0   1.0   221  0.096930
15   2.0   0.0   204  0.089474
1    0.0   1.0   173  0.075877
0    0.0   0.0   171  0.075000
16   2.0   1.0   171  0.075000
9    1.0   2.0   156  0.068421
17   2.0   2.0   119  0.052193
23   3.0   1.0   112  0.049123
22   3.0   0.0    96  0.042105